_episodes/03-monitor-the-model.md

```diff
@@ -0,0 +1,724 @@
+This can in principle also be done for deep learning.
+An alternative, more common approach, is to add **BatchNormalization** layers which will learn how to scale the input values.
+Similar to dropout, batch normalization is available as a network layer in keras and can be added to the network in a similar way.
+It does not require any additional parameter setting. 
+
+~~~
+from tensorflow.keras.layers import BatchNormalization
+~~~
+{: .language-python} 
+
+## Exercise: Add a BatchNormalization layer as the first layer to your neural network.
+(documentation & reference: https://keras.io/api/layers/normalization_layers/batch_normalization/)
+
+~~~
+def create_nn(n_features, n_predictions):
+    # Input layer
+    layers_input = Input(shape=(n_features,), name='input')
+
+    # Dense layers
+    layers_dense = BatchNormalization()(layers_input)
+    layers_dense = Dense(100, 'relu')(layers_dense)
+    layers_dense = Dropout(rate=0.2)(layers_dense)
+    layers_dense = Dense(50, 'relu')(layers_dense)
+    layers_dense = Dropout(rate=0.2)(layers_dense)
+
+    # Output layer
+    layers_output = Dense(n_predictions)(layers_dense)
+
+    # Defining the model and compiling it
+    return Model(inputs=layers_input, outputs=layers_output, name="model_dropout_batchnorm")
+
+model = create_nn(X_data.shape[1], 1)
+model.compile(loss='mse', optimizer=Adam(1e-4), metrics=[tf.keras.metrics.RootMeanSquaredError()])
+model.summary()
+~~~
+{: .language-python}      
+
+Which is then trained as above:
+~~~
+history = model.fit(X_train, y_train,
+                    batch_size = 50,
+                    epochs = 1000,
+                    validation_data=(X_val, y_val),
+                    callbacks=[earlystopper],
+                    verbose = 2)
+
+history_df = pd.DataFrame.from_dict(history.history)
+sns.lineplot(data=history_df[['root_mean_squared_error', 'val_root_mean_squared_error']])
+plt.xlabel("epochs")
+plt.ylabel("RMSE")
+~~~
+{: .language-python}      
+
+![Output of plotting sample](../fig/03_training_history_5_rmse_batchnorm.png)
+
+## Run on test set and compare to naive baseline
+It seems that no matter what we add, the overall loss does not decrease much further (we at least avoided overfitting though!).
+Let's again plot the results on the test set:
+~~~
+y_test_predicted = model.predict(X_test)
+
+plt.figure(figsize=(5, 5), dpi=100)
+plt.scatter(y_test_predicted, y_test, s=10, alpha=0.5)
+plt.xlabel("predicted sunshine hours")
+plt.ylabel("true sunshine hours")
+~~~
+{: .language-python} 
+
+![Output of plotting sample](../fig/03_regression_test_5_dropout_batchnorm.png)
+
+Well... certainly not perfect. But how good or bad is this? Maybe not good enough to plan your picnic for tomorrow.
+But let's better compare it to a naive baseline.
+Maybe the simplest prediction to make would be to say: Tomorrow we will have the same number of sunshine hours as today. Let's compare to this.
+
+~~~
```

Comment:
I would make an exercise of this. Ask them to plot the baseline, and calculate the RMSE (probably need to refer to https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_squared_error.html ) to compare